### PR TITLE
Adds FillNodes workload for mlmd_bench.

### PR DESCRIPTION
**Pull request by BrianSong**
For now, the FillNodes workload only support insert.
Will support update in near feature.

### CODE REVIEW

File: ml_metadata/tools/mlmd_bench/util.h [hunk 1]
```diff
@@ -20,21 +20,42 @@ limitations under the License.
...
 #include "ml_metadata/proto/metadata_store.pb.h"
-#include "ml_metadata/tools/mlmd_bench/proto/mlmd_bench.pb.h"
 #include "tensorflow/core/lib/core/status.h"
 
 namespace ml_metadata {
 
-// Defines a Type can be ArtifactType / ExecutionType / ContextType.
-using Type = ::absl::variant<ArtifactType, ExecutionType, ContextType>;
+// Defines a Type that can be ArtifactType / ExecutionType / ContextType.
+using Type = absl::variant<ArtifactType, ExecutionType, ContextType>;
```
Comment (hughmiao) [inline ml_metadata/tools/mlmd_bench/util.h]: using ::asbl  when [1] defining type names

[1] https://google.github.io/styleguide/cppguide.html#Namespaces

File: ml_metadata/tools/mlmd_bench/util.h [hunk 2]
```diff
@@ -20,21 +20,42 @@ limitations under the License.
...
 
-// Gets all the existing types (the specific types that indicated by
-// `fill_types_config`) inside db and store them into `existing_types`.
-// Returns detailed error if query executions failed.
-tensorflow::Status GetExistingTypes(const FillTypesConfig& fill_types_config,
-                                    MetadataStore* store,
+// Defines a NodeType that can be Artifact / Execution / Context.
+using NodeType = absl::variant<Artifact, Execution, Context>;
```
Comment (hughmiao) [inline ml_metadata/tools/mlmd_bench/util.h]: ditto, nit `Node` instead of `NodeType`

File: ml_metadata/tools/mlmd_bench/util.h [hunk 3]
```diff
@@ -20,21 +20,42 @@ limitations under the License.
...
+
+// Gets all the existing types inside db and store them into
+// `existing_types`. The specific type for the types will be indicated by
+// `specification`. Returns detailed error if query executions failed.
+tensorflow::Status GetExistingTypes(int specification, MetadataStore* store,
```
Comment (hughmiao) [inline ml_metadata/tools/mlmd_bench/util.h]: let's pass MetadataStore&  for input parameters, ditto for the other MetadataStore*

File: ml_metadata/tools/mlmd_bench/fill_nodes_workload.h [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+#include "ml_metadata/tools/mlmd_bench/proto/mlmd_bench.pb.h"
+#include "ml_metadata/tools/mlmd_bench/workload.h"
+#include "tensorflow/core/lib/core/status.h"
+
+namespace ml_metadata {
+
+// Defines a FillNodesWorkItemType that can be PutArtifactsRequest /
+// PutExecutionsRequest / PutContextsRequest.
+using FillNodesWorkItemType =
+    absl::variant<PutArtifactsRequest, PutExecutionsRequest,
```
Comment (hughmiao) [inline ml_metadata/tools/mlmd_bench/fill_nodes_workload.h]: ditto

File: ml_metadata/tools/mlmd_bench/fill_types_workload.cc [hunk 5]
```diff
@@ -190,8 +190,8 @@ tensorflow::Status FillTypes::SetUpImpl(MetadataStore* store) {
 
   // Gets all the existing current types inside db for later update cases.
   std::vector<Type> existing_types;
-  TF_RETURN_IF_ERROR(
-      GetExistingTypes(fill_types_config_, store, existing_types));
+  TF_RETURN_IF_ERROR(GetExistingTypes(fill_types_config_.specification(), store,
```
Comment (hughmiao) [inline ml_metadata/tools/mlmd_bench/fill_types_workload.cc]: update the *store

File: ml_metadata/tools/mlmd_bench/fill_types_workload_test.cc [hunk 6]
```diff
@@ -184,8 +153,8 @@ TEST_P(FillTypesInsertParameterizedTestFixture, InsertWhenDbIsEmptyTest) {
   TF_ASSERT_OK(ExecuteWorkload(store_.get(), fill_types_));
   // Gets all the existing current types inside db after the insert.
   std::vector<Type> existing_types;
-  TF_ASSERT_OK(GetExistingTypes(GetParam().fill_types_config(), store_.get(),
-                                existing_types));
+  TF_ASSERT_OK(GetExistingTypes(GetParam().fill_types_config().specification(),
+                                store_.get(), existing_types));
```
Comment (hughmiao) [inline ml_metadata/tools/mlmd_bench/fill_types_workload_test.cc]: *store_

File: ml_metadata/tools/mlmd_bench/util.h [hunk 7]
```diff
@@ -20,21 +20,42 @@ limitations under the License.
...
 
+// Gets all the existing nodes inside db and store them into `existing_nodes`.
+// The specific type for the nodes will be indicated by `specification`. Returns
+// detailed error if query executions failed.
+tensorflow::Status GetExistingNodes(int specification, MetadataStore* store,
+                                    std::vector<NodeType>& existing_nodes);
+
+// Inserts some types into db for setting up in testing. Returns detailed error
+// if query executions failed.
+tensorflow::Status InsertTypesInDb(int64 num_artifact_types,
```
Comment (hughmiao) [inline ml_metadata/tools/mlmd_bench/util.h:56]: add some simple test

File: ml_metadata/tools/mlmd_bench/util.cc [hunk 8]
```diff
@@ -58,8 +84,137 @@ tensorflow::Status GetExistingTypes(const FillTypesConfig& fill_types_config,
...
+  GetExistingTypes(/*specification=*/1, store, existing_execution_types);
+  std::vector<Type> existing_context_types;
+  GetExistingTypes(/*specification=*/2, store, existing_context_types);
+
+  std::uniform_int_distribution<int64> uniform_dist_artifact_type_index{
+      0, (int64)(existing_artifact_types.size() - 1)};
+  std::uniform_int_distribution<int64> uniform_dist_execution_type_index{
+      0, (int64)(existing_execution_types.size() - 1)};
+  std::uniform_int_distribution<int64> uniform_dist_context_type_index{
+      0, (int64)(existing_context_types.size() - 1)};
```
Comment (hughmiao) [inline ml_metadata/tools/mlmd_bench/util.cc]: drop random flavor in tests, if random using a fixed seed.

File: ml_metadata/tools/mlmd_bench/util.cc [hunk 9]
```diff
@@ -58,8 +84,137 @@ tensorflow::Status GetExistingTypes(const FillTypesConfig& fill_types_config,
...
+
+  for (int64 i = 0; i < num_artifact_nodes; i++) {
+    const int64 type_index = uniform_dist_artifact_type_index(gen);
+    const string& node_name = absl::StrCat("pre_insert_artifact-", i);
+    PutArtifactsRequest put_request;
+    Artifact* curr_node = put_request.add_artifacts();
+    PrepareNode<ArtifactType, Artifact>(
+        node_name, absl::get<ArtifactType>(existing_artifact_types[type_index]),
+        *curr_node);
+    SetArtifactAdditionalFields(node_name, *curr_node);
```
Comment (hughmiao) [inline ml_metadata/tools/mlmd_bench/util.cc]: Artifact a0;
a0.set_uri("xxx")
for (int64 i = 0; i< ... ) {
    Artifact ai = a0;
    a0.set_name(i) ..
    a0.set_type (i..)
    a0.set_uri ...(i) ..
    PrepareNode(a0) ..
}

File: ml_metadata/tools/mlmd_bench/util.cc [hunk 10]
```diff
@@ -58,8 +84,137 @@ tensorflow::Status GetExistingTypes(const FillTypesConfig& fill_types_config,
...
+  for (int64 i = 0; i < num_execution_nodes; i++) {
+    const int64 type_index = uniform_dist_execution_type_index(gen);
+    const string& node_name = absl::StrCat("pre_insert_execution-", i);
+    PutExecutionsRequest put_request;
+    Execution* curr_node = put_request.add_executions();
+    PrepareNode<ExecutionType, Execution>(
+        node_name,
+        absl::get<ExecutionType>(existing_execution_types[type_index]),
+        *curr_node);
+    SetExecutionAdditionalFields(*curr_node);
```
Comment (hughmiao) [inline ml_metadata/tools/mlmd_bench/util.cc]: ditto

File: ml_metadata/tools/mlmd_bench/util.cc [hunk 11]
```diff
@@ -58,8 +84,137 @@ tensorflow::Status GetExistingTypes(const FillTypesConfig& fill_types_config,
...
+
+  return store->PutTypes(put_request, &put_response);
+}
+
+tensorflow::Status InsertNodesInDb(const int64 num_artifact_nodes,
+                                   const int64 num_execution_nodes,
+                                   const int64 num_context_nodes,
+                                   MetadataStore* store) {
+  std::vector<Type> existing_artifact_types;
+  GetExistingTypes(/*specification=*/0, store, existing_artifact_types);
```
Comment (hughmiao) [inline ml_metadata/tools/mlmd_bench/util.cc]: TF_RETURN_IF_ERROR

File: ml_metadata/tools/mlmd_bench/util.cc [hunk 12]
```diff
@@ -14,23 +14,47 @@ limitations under the License.
...
+// Sets additional fields(uri, state) for artifacts.
+void SetArtifactAdditionalFields(const string& node_name, Artifact& node) {
+  node.set_uri(absl::StrCat(node_name, "_uri"));
+  node.set_state(Artifact::UNKNOWN);
+}
+
+// Sets additional field(state) for executions.
+void SetExecutionAdditionalFields(Execution& node) {
+  node.set_last_known_state(Execution::UNKNOWN);
+}
```
Comment (hughmiao) [inline ml_metadata/tools/mlmd_bench/util.cc]: drop

File: ml_metadata/tools/mlmd_bench/util.cc [hunk 13]
```diff
@@ -14,23 +14,47 @@ limitations under the License.
...
+namespace {
 
-tensorflow::Status GetExistingTypes(const FillTypesConfig& fill_types_config,
+template <typename T, typename NT>
+void PrepareNode(const std::string& node_name, const T& curr_type,
+                 NT& curr_node) {
+  curr_node.set_type_id(curr_type.id());
+  curr_node.set_name(node_name);
+  (*curr_node.mutable_properties())["property"].set_string_value("*");
+}
```
Comment (hughmiao) [inline ml_metadata/tools/mlmd_bench/util.cc]: + mutable_custom_properties ..

File: ml_metadata/tools/mlmd_bench/util.cc [hunk 14]
```diff
@@ -14,23 +14,47 @@ limitations under the License.
...
 namespace ml_metadata {
+namespace {
 
-tensorflow::Status GetExistingTypes(const FillTypesConfig& fill_types_config,
+template <typename T, typename NT>
+void PrepareNode(const std::string& node_name, const T& curr_type,
+                 NT& curr_node) {
+  curr_node.set_type_id(curr_type.id());
+  curr_node.set_name(node_name);
+  (*curr_node.mutable_properties())["property"].set_string_value("*");
```
Comment (hughmiao) [inline ml_metadata/tools/mlmd_bench/util.cc]: nit, * -> 'foo'/'bar'/'baz'

File: ml_metadata/tools/mlmd_bench/util.cc [hunk 15]
```diff
@@ -14,23 +14,47 @@ limitations under the License.
...
+}  // namespace
+
+tensorflow::Status GetExistingTypes(const int specification,
                                     MetadataStore* store,
                                     std::vector<Type>& existing_types) {
-  switch (fill_types_config.specification()) {
-    case FillTypesConfig::ARTIFACT_TYPE: {
+  switch (specification) {
+    // Gets ArtifactTypes.
+    case 0: {
```
Comment (hughmiao) [inline ml_metadata/tools/mlmd_bench/util.cc]: option 1:
GetExistingTypes(absl::variant<FillTypeConfig, FillNodeConfig> config ..,
                               store, existing_types)

option 2:
template <T>
2.1. GetExistingTypes(store, existing_types)
2.2. GetExistingTypes(store, T& msg, existing_types)

if (is_same<Type, ArtifactType>::value ..) {

} else if (is_same<Type, ExecutionType>::value ..) {

}

File: ml_metadata/tools/mlmd_bench/fill_nodes_workload.cc [hunk 16]
```diff
@@ -0,0 +1,321 @@
...
+  switch (fill_nodes_config_.specification()) {
+    case FillNodesConfig::ARTIFACT: {
+      name_ = "fill_artifact";
+      break;
+    }
+    case FillNodesConfig::EXECUTION: {
+      name_ = "fill_execution";
+      break;
+    }
+    case FillNodesConfig::CONTEXT: {
```
Comment (hughmiao) [inline ml_metadata/tools/mlmd_bench/fill_nodes_workload.cc]: qq, fill_nodes_config_.specification().DebugString() ?

absl::StrCat("fill_", fill_nodes_config_.specification().DebugString() )

enum string value functions [1]

[1] https://developers.google.com/protocol-buffers/docs/reference/cpp-generated#enum

File: ml_metadata/tools/mlmd_bench/fill_nodes_workload.cc [hunk 17]
```diff
@@ -0,0 +1,321 @@
...
+  if (fill_nodes_config_.update()) {
+    name_ += "(update)";
+  }
+}
+
+tensorflow::Status FillNodes::SetUpImpl(MetadataStore* store) {
+  LOG(INFO) << "Setting up ...";
+
+  // Gets all the specific types in db to choose from when generating nodes.
+  std::vector<Type> existing_types;
```
Comment (hughmiao) [inline ml_metadata/tools/mlmd_bench/fill_nodes_workload.cc:220]: + // If there's no types in the store, return error.

File: ml_metadata/tools/mlmd_bench/fill_nodes_workload.cc [hunk 18]
```diff
@@ -0,0 +1,321 @@
...
+  TF_RETURN_IF_ERROR(GetAndValidateExistingTypes(
+      fill_nodes_config_.specification(), store, existing_types));
+  std::uniform_int_distribution<int64> uniform_dist_type_index{
+      0, (int64)(existing_types.size() - 1)};
+
+  std::minstd_rand0 gen(absl::ToUnixMillis(absl::Now()));
+
+  std::string node_batch_name;
+  int64 curr_bytes, num_properties, string_value_bytes, num_nodes, type_index;
```
Comment (hughmiao) [inline ml_metadata/tools/mlmd_bench/fill_nodes_workload.cc]: move to line 238

struct NodeParam {
    curr_bytes .
    num ...
    string ..
    ..
}

for .. {
    NodeParam param;
}

File: ml_metadata/tools/mlmd_bench/fill_nodes_workload.cc [hunk 19]
```diff
@@ -0,0 +1,321 @@
...
+    curr_bytes = 0;
+    FillNodesWorkItemType put_request;
+    InitializeCurrentNodeBatchParameters(
+        i, fill_nodes_config_, uniform_dist_type_index, gen, node_batch_name,
+        num_properties, string_value_bytes, num_nodes, type_index);
+    switch (fill_nodes_config_.specification()) {
+      case FillNodesConfig::ARTIFACT: {
+        InitializePutRequest<PutArtifactsRequest>(num_nodes, put_request);
+        google::protobuf::RepeatedPtrField<Artifact>& node_batch =
+            *absl::get<PutArtifactsRequest>(put_request).mutable_artifacts();
```
Comment (hughmiao) [inline ml_metadata/tools/mlmd_bench/fill_nodes_workload.cc]: auto nodes = absl::get<PutArtifactsRequests>(put_request).mutable_artifacts();

*nodes ..

[1] https://google.github.io/styleguide/cppguide.html#Type_deduction

File: ml_metadata/tools/mlmd_bench/fill_nodes_workload.cc [hunk 20]
```diff
@@ -0,0 +1,321 @@
...
+
+// Executions of work items.
+tensorflow::Status FillNodes::RunOpImpl(const int64 work_items_index,
+                                        MetadataStore* store) {
+  switch (fill_nodes_config_.specification()) {
+    case FillNodesConfig::ARTIFACT: {
+      PutArtifactsRequest put_request =
+          absl::get<PutArtifactsRequest>(work_items_[work_items_index].first);
+      PutArtifactsResponse put_response;
+      TF_RETURN_IF_ERROR(store->PutArtifacts(put_request, &put_response));
```
Comment (hughmiao) [inline ml_metadata/tools/mlmd_bench/fill_nodes_workload.cc]: return store->PutArt..

File: ml_metadata/tools/mlmd_bench/fill_nodes_workload.cc [hunk 21]
```diff
@@ -0,0 +1,321 @@
...
+           GetTransferredBytesForNodeProperties(node.custom_properties());
+  return bytes;
+}
+
+// Generates insert node.
+// For insert cases, the node's type will be `type` and its properties will be
+// generated w.r.t. `node_name`, `num_properties` and `string_value_bytes`.
+// Leaves it as return `void` instead of return `int64` for now because
+// FillNodes update mode will return `tensorflow::Status` in the future.
+template <typename T, typename NT>
```
Comment (hughmiao) [inline ml_metadata/tools/mlmd_bench/fill_nodes_workload.cc]: template <typename Type, typename Node> T/N

File: ml_metadata/tools/mlmd_bench/fill_nodes_workload.cc [hunk 22]
```diff
@@ -0,0 +1,321 @@
...
+  // Insert nodes cases.
+  // Loops over all the node inside `node_batch` and sets up one by one.
+  for (int64 i = 0; i < node_batch.size(); ++i) {
+    node_batch[i].set_name(absl::StrCat(node_batch_name, "_node_", i));
+    node_batch[i].set_type_id(type.id());
+    // Uses "********" as the fake property value for current node.
+    std::string property_value(string_value_bytes, '*');
+    int64 curr_num_properties = 0;
+    // Loops over the types properties while generating the node's properties
+    // accordingly.
```
Comment (hughmiao) [inline ml_metadata/tools/mlmd_bench/fill_nodes_workload.cc:150]: int populated_num_properties = 0;
for (auto p : type.properties()) {
   // get p name, get p's value type  
   set_string_value / set_int_value / set_float_value
   node_batch[i].mutable_properties()[name].set...(...)
   if (++populated_num_properties > ..) {
       break;
   }
}

leave a TODO looks ok for now.

File: ml_metadata/tools/mlmd_bench/fill_nodes_workload_test.cc [hunk 23]
```diff
@@ -0,0 +1,192 @@
...
+constexpr int kNumberOfOperations = 100;
+constexpr int kNumberOfExistedTypesInDb = 100;
+constexpr int kNumberOfExistedNodesInDb = 100;
+constexpr int kNumberOfNodesPerRequest = 10;
+
+// Enumerates the workload configurations as the test parameters that ensure
+// test coverage.
+std::vector<WorkloadConfig> EnumerateConfigs(const bool is_update) {
+  std::vector<WorkloadConfig> config_vector;
+  WorkloadConfig template_config = testing::ParseTextProtoOrDie<WorkloadConfig>(
```
Comment (hughmiao) [inline ml_metadata/tools/mlmd_bench/fill_nodes_workload_test.cc]: constexpr absl::string_view config_str = R"(. ..");

{
WorkloadConfig config = ParseTextProto..(config_str);
confi.mutable_fill_nod
}

{
WorkloadConfig config = ParseTextProto..(config_str);
confi.mutable_fill_nod
}

{
WorkloadConfig config = ParseTextProto..(config_str);
confi.mutable_fill_nod
}

**Review by hughmiao — CHANGES_REQUESTED** [review]
thanks, Brian! some comments.

File: ml_metadata/tools/mlmd_bench/util.h (hunk 3 above)
Comment (BrianSong) [reply ml_metadata/tools/mlmd_bench/util.h]: Done. For now, I only fixed the MetadataStore* in the `util.h` and its related file as we discussed.

File: ml_metadata/tools/mlmd_bench/fill_nodes_workload.cc (hunk 16 above)
Comment (BrianSong) [reply ml_metadata/tools/mlmd_bench/fill_nodes_workload.cc]: `DebugString()` can only used for message interface.
Used `Specification_Name()` instead.

File: ml_metadata/tools/mlmd_bench/proto/mlmd_bench.proto [hunk 24]
```diff
@@ -51,13 +51,36 @@ message FillTypesConfig {
...
+
 // The mlmd_bench workload config.
 message WorkloadConfig {
   oneof workload_config {
     FillTypesConfig fill_types_config = 1;
+    FillNodesConfig fill_nodes_config = 2;
   }
   // The number of operations to be run in parallel.
-  optional int64 num_operations = 2;
+  optional int64 num_operations = 3;
```
Comment (hughmiao) [inline ml_metadata/tools/mlmd_bench/proto/mlmd_bench.proto]: = 2;

File: ml_metadata/tools/mlmd_bench/proto/mlmd_bench.proto [hunk 25]
```diff
@@ -51,13 +51,36 @@ message FillTypesConfig {
...
+  // Specifies the number of nodes to be filled per request, 
+  // modeled by a uniform distribution.
+  optional UniformDistribution num_nodes = 5;
+}
+
 // The mlmd_bench workload config.
 message WorkloadConfig {
   oneof workload_config {
     FillTypesConfig fill_types_config = 1;
+    FillNodesConfig fill_nodes_config = 2;
```
Comment (hughmiao) [inline ml_metadata/tools/mlmd_bench/proto/mlmd_bench.proto]: = 3;

File: ml_metadata/tools/mlmd_bench/util.cc [hunk 26]
```diff
@@ -25,42 +26,239 @@ limitations under the License.
...
-  switch (fill_types_config.specification()) {
-    case FillTypesConfig::ARTIFACT_TYPE: {
+// Enumeration type that indicates which node type to be fetched using
+// GetExistingTypesImpl().
+enum FetchType { FetchArtifactType, FetchExecutionType, FetchContextType };
+
+// Prepares node's properties for inserting inside InsertNodesInDb().
+template <typename T, typename NT>
+void PrepareNode(const std::string& node_name, const T& curr_type,
+                 NT& curr_node) {
```
Comment (hughmiao) [inline ml_metadata/tools/mlmd_bench/util.cc]: nit, s/curr_node/node

File: ml_metadata/tools/mlmd_bench/util.cc [hunk 27]
```diff
@@ -25,42 +26,239 @@ limitations under the License.
...
+  return tensorflow::Status::OK();
+}
+
+tensorflow::Status InsertTypesInDb(const int64 num_artifact_types,
+                                   const int64 num_execution_types,
+                                   const int64 num_context_types,
+                                   MetadataStore& store) {
+  PutTypesRequest put_request;
+  PutTypesResponse put_response;
+
```
Comment (hughmiao) [inline ml_metadata/tools/mlmd_bench/util.cc:175]: move FormatTime(absl::Now()) here

File: ml_metadata/tools/mlmd_bench/util.cc [hunk 28]
```diff
@@ -25,42 +26,239 @@ limitations under the License.
...
+      GetExistingTypes(fill_types_config, store, existing_execution_types));
+  fill_types_config.set_specification(FillTypesConfig::CONTEXT_TYPE);
+  std::vector<Type> existing_context_types;
+  TF_RETURN_IF_ERROR(
+      GetExistingTypes(fill_types_config, store, existing_context_types));
+
+  for (int64 i = 0; i < num_artifact_nodes; i++) {
+    const int64 type_index = i % existing_artifact_types.size();
+    const string& node_name = absl::StrCat(
+        "pre_insert_artifact-", absl::FormatTime(absl::Now()), "-", i);
```
Comment (hughmiao) [inline ml_metadata/tools/mlmd_bench/util.cc]: move Now() ouside of the loop

File: ml_metadata/tools/mlmd_bench/util.cc [hunk 29]
```diff
@@ -25,42 +26,239 @@ limitations under the License.
...
+    PutArtifactsRequest put_request;
+    Artifact* curr_node = put_request.add_artifacts();
+    PrepareNode<ArtifactType, Artifact>(
+        node_name, absl::get<ArtifactType>(existing_artifact_types[type_index]),
+        *curr_node);
+    curr_node->set_uri(absl::StrCat(node_name, "_uri"));
+    curr_node->set_state(Artifact::UNKNOWN);
+    PutArtifactsResponse put_response;
+    TF_RETURN_IF_ERROR(store.PutArtifacts(put_request, &put_response));
```
Comment (hughmiao) [inline ml_metadata/tools/mlmd_bench/util.cc]: move PutArtifacts after the loop

File: ml_metadata/tools/mlmd_bench/util_test.cc [hunk 30]
```diff
@@ -0,0 +1,219 @@
...
+  TF_ASSERT_OK(store->GetArtifactTypes(
+      /*request=*/{}, &get_artifact_types_response));
+  GetExecutionTypesResponse get_execution_types_response;
+  TF_ASSERT_OK(store->GetExecutionTypes(
+      /*request=*/{}, &get_execution_types_response));
+  GetContextTypesResponse get_context_types_response;
+  TF_ASSERT_OK(store->GetContextTypes(
+      /*request=*/{}, &get_context_types_response));
+
+  ASSERT_EQ(kNumberOfInsertedArtifactTypes,
```
Comment (hughmiao) [inline ml_metadata/tools/mlmd_bench/util_test.cc]: EXPECT_EQ

better, EXPECT_THAT(get_artifact_types_response.artifact_types(), ::testing::SizeIs(kNumber ...))

File: ml_metadata/tools/mlmd_bench/util_test.cc [hunk 31]
```diff
@@ -0,0 +1,219 @@
...
+}
+
+// Tests InsertNodesInDb().
+TEST(UtilInsertTest, InsertNodesTest) {
+  std::unique_ptr<MetadataStore> store;
+  ConnectionConfig mlmd_config;
+  // Uses a fake in-memory SQLite database for testing.
+  mlmd_config.mutable_fake_database();
+  TF_ASSERT_OK(CreateMetadataStore(mlmd_config, &store));
+  // InsertTypesInDb() has passed the tests.
```
Comment (hughmiao) [inline ml_metadata/tools/mlmd_bench/util_test.cc]: drop

File: ml_metadata/tools/mlmd_bench/util_test.cc [hunk 32]
```diff
@@ -0,0 +1,219 @@
...
+  TF_ASSERT_OK(store->GetArtifacts(
+      /*request=*/{}, &get_artifacts_response));
+  GetExecutionsResponse get_executions_response;
+  TF_ASSERT_OK(store->GetExecutions(
+      /*request=*/{}, &get_executions_response));
+  GetContextsResponse get_contexts_response;
+  TF_ASSERT_OK(store->GetContexts(
+      /*request=*/{}, &get_contexts_response));
+
+  ASSERT_EQ(kNumberOfInsertedArtifacts,
```
Comment (hughmiao) [inline ml_metadata/tools/mlmd_bench/util_test.cc]: ditto

File: ml_metadata/tools/mlmd_bench/util_test.cc [hunk 33]
```diff
@@ -0,0 +1,219 @@
...
+    TF_ASSERT_OK(GetExistingTypes(fill_types_config, *store, exisiting_types));
+    EXPECT_EQ(kNumberOfInsertedArtifactTypes, exisiting_types.size());
+  }
+
+  {
+    std::vector<Type> exisiting_types;
+    FillTypesConfig fill_types_config;
+    fill_types_config.set_specification(FillTypesConfig::EXECUTION_TYPE);
+    TF_ASSERT_OK(GetExistingTypes(fill_types_config, *store, exisiting_types));
+    EXPECT_EQ(kNumberOfInsertedExecutionTypes, exisiting_types.size());
```
Comment (hughmiao) [inline ml_metadata/tools/mlmd_bench/util_test.cc]: ditto, SizeIs

File: ml_metadata/tools/mlmd_bench/fill_types_workload_test.cc [hunk 34]
```diff
@@ -36,59 +36,46 @@ constexpr int kNumberOfExisitedTypeInInsert = 80;
...
+    config.set_num_operations(kNumberOfOperations);
+    config.mutable_fill_types_config()->set_update(is_update);
+    config.mutable_fill_types_config()->set_specification(
+        FillTypesConfig::CONTEXT_TYPE);
+    config_vector.push_back(config);
   }
 
-  TF_RETURN_IF_ERROR(store->PutTypes(put_request, &put_response));
-  return tensorflow::Status::OK();
+  return config_vector;
```
Comment (hughmiao) [inline ml_metadata/tools/mlmd_bench/fill_types_workload_test.cc]: nit, configs

File: ml_metadata/tools/mlmd_bench/BUILD [hunk 35]
```diff
@@ -160,6 +201,23 @@ cc_library(
     ],
 )
 
+ml_metadata_cc_test(
+    name = "util_test",
+    size = "small",
+    srcs = ["util_test.cc"],
+    deps = [
+        ":util",
+        #"@com_google_googletest//:gtest",
```
Comment (hughmiao) [inline ml_metadata/tools/mlmd_bench/BUILD]: sync from head?

File: ml_metadata/tools/mlmd_bench/fill_nodes_workload.h [hunk 36]
```diff
@@ -0,0 +1,79 @@
...
+  // workload.
+  std::string GetName() final;
+
+ private:
+  // Workload configurations specified by the users.
+  const FillNodesConfig fill_nodes_config_;
+  // Number of operations for the current workload.
+  const int64 num_operations_;
+  // String for indicating the name of current workload instance.
+  std::string name_;
```
Comment (hughmiao) [inline ml_metadata/tools/mlmd_bench/fill_nodes_workload.h]: const?

File: ml_metadata/tools/mlmd_bench/fill_nodes_workload.cc [hunk 37]
```diff
@@ -0,0 +1,305 @@
...
+
+tensorflow::Status FillNodes::SetUpImpl(MetadataStore* store) {
+  LOG(INFO) << "Setting up ...";
+  int64 curr_bytes = 0;
+
+  // Gets all the specific types in db to choose from when generating nodes.
+  // If there's no types in the store, return error.
+  std::vector<Type> existing_types;
+  TF_RETURN_IF_ERROR(
+      GetAndValidateExistingTypes(fill_nodes_config_, store, existing_types));
```
Comment (hughmiao) [inline ml_metadata/tools/mlmd_bench/fill_nodes_workload.cc]: *store

File: ml_metadata/tools/mlmd_bench/fill_nodes_workload.cc [hunk 38]
```diff
@@ -0,0 +1,305 @@
...
+  std::uniform_int_distribution<int64> uniform_dist_type_index{
+      0, (int64)(existing_types.size() - 1)};
```
Comment (hughmiao) [inline ml_metadata/tools/mlmd_bench/fill_nodes_workload.cc:224]: double check [a, b) or [a, b]

**Review by hughmiao — APPROVED** [review]
some nit comments, please fix them, and wait for the pending cls and sync to head before sending public pr.

File: ml_metadata/tools/mlmd_bench/fill_nodes_workload.cc (hunk 38 above)
Comment (BrianSong) [reply ml_metadata/tools/mlmd_bench/fill_nodes_workload.cc:224]: Yes, I checked. It is [a, b] :)

File: ml_metadata/tools/mlmd_bench/fill_nodes_workload.h (hunk 36 above)
Comment (BrianSong) [reply ml_metadata/tools/mlmd_bench/fill_nodes_workload.h]: Fixed! Added lambda function inside initialization list for the `name_`. Thanks!

**Pull request merged**